javascript/ql/src/semmle/javascript/frameworks/Electron.qll

```diff
@@ -6,31 +6,466 @@ module Electron {
    */
   class WebPreferences extends DataFlow::ObjectLiteralNode {
     WebPreferences() {
-      exists(BrowserObject bo | this = bo.getOptionArgument(0, "webPreferences").getALocalSource())
+      exists(NewBrowserObject bo | this = bo.getWebPreferences().getALocalSource())
     }
   }
   
+  /**
+   * a data flow node that is an Electron `BrowserView` or `BrowserWindow`
+   */
+  abstract class BrowserObject extends DataFlow::TrackedNode {
+  }
+  
   /**
    * A data flow node that creates a new `BrowserWindow` or `BrowserView`.
    */
-  private abstract class BrowserObject extends DataFlow::NewNode {
+  abstract class NewBrowserObject extends BrowserObject {
+    NewBrowserObject() {
+      this instanceof DataFlow::NewNode
+    }
+    
+    DataFlow::Node getWebPreferences() {
+      result = this.(DataFlow::NewNode).getOptionArgument(0, "webPreferences")
+    }
   }
   
   /**
    * A data flow node that creates a new `BrowserWindow`.
    */
-  class BrowserWindow extends BrowserObject {
+  class BrowserWindow extends NewBrowserObject {
     BrowserWindow() {
       this = DataFlow::moduleMember("electron", "BrowserWindow").getAnInstantiation()
     }
   }
   
+  /**
+   * A data flow node with a TypeScript type indicating it is an Electron `BrowserWindow`
+   */
+  class TypedBrowserWindow extends BrowserObject {
+    TypedBrowserWindow() {
+      this.asExpr().getType().toString() = "BrowserWindow"
+    }
+  }
+  
+  /**
+   * A data flow node with a TypeScript type indicating it is an Electron `BrowserView`
+   */
+  class TypedBrowserView extends BrowserObject {
+    TypedBrowserView() {
+      this.asExpr().getType().toString() = "BrowserView"
+    }
+  }
   /**
    * A data flow node that creates a new `BrowserView`.
    */
-  class BrowserView extends BrowserObject {
+  class BrowserView extends NewBrowserObject {
     BrowserView() {
       this = DataFlow::moduleMember("electron", "BrowserView").getAnInstantiation()
     }
   }
-}
\ No newline at end of file
+  
+  /**
+   * A data flow node that is the `webContents` property of an Electron browser object
+   */
+  cached class WebContents extends DataFlow::TrackedNode {
+    cached WebContents() {
+      exists(BrowserObject bo |
+        bo.flowsTo(this.(DataFlow::PropRead).getBase())
+      )
+    }
+  }
+  
+  /**
+   * A data flow node that imports `electron.ipcMain`.
+   */
+  class IPCMain extends DataFlow::SourceNode {
```

Comment:
Does this need to be public?

Reply:
I've been talking with a customer about their uses of Electron IPC, and it looks like we'll need to write custom queries against this library, so I'd prefer to keep these public. Would putting them in a submodule (`Electron::IPC` or `Electron::ElectronIPC`, maybe) be a good compromise? The new `AdditionalFlowStep`s would still be active, the message and callback classes would be available for use in custom queries, and the namespace wouldn't be too polluted.